consensus/src/chained_bft/block_storage/pending_votes_test.rs

```diff
@@ -0,0 +1,294 @@
+
+fn random_ledger_info() -> LedgerInfo {
+    LedgerInfo::new(
+        0,
+        HashValue::random(),
+        HashValue::random(),
+        HashValue::random(),
+        0,
+        0,
+        None,
+    )
+}
+
+fn random_vote_data(round: Round) -> VoteData {
+    assert!(round >= 1);
+    VoteData::new(
+        HashValue::random(),
+        HashValue::random(),
+        round,
+        HashValue::random(),
+        round - 1,
+        HashValue::random(),
+        if round >= 2 { round - 2 } else { round - 1 },
+    )
+}
+
+#[test]
+/// Verify that votes are properly aggregated to QC based on their LedgerInfo digest
+fn test_qc_aggregation() {
+    ::logger::try_init_for_testing();
+    let (signers, validator) = random_validator_verifier(4, Some(2), false);
+    let validator_verifier = Arc::new(validator);
+    let mut pending_votes = PendingVotes::new();
+
+    let li1 = random_ledger_info();
+    let vote_data_1 = random_vote_data(1);
+    let vote_data_1_author_0 = VoteMsg::new(
+        vote_data_1.clone(),
+        signers[0].author(),
+        li1.clone(),
+        &signers[0],
+    );
+
+    // first time a new vote is added the result is VoteAdded
+    assert_eq!(
+        pending_votes.insert_vote(&vote_data_1_author_0, Arc::clone(&validator_verifier)),
+        VoteReceptionResult::VoteAdded(1)
+    );
+    // same author voting for the same thing: result is DuplicateVote
+    assert_eq!(
+        pending_votes.insert_vote(&vote_data_1_author_0, Arc::clone(&validator_verifier)),
+        VoteReceptionResult::DuplicateVote
+    );
+    // same author voting for a different result in the same round:
+    // override the prev value and return equivocation
+    let li2 = random_ledger_info();
+    let vote_data_2 = random_vote_data(1);
+    let vote_data_2_author_0 = VoteMsg::new(
+        vote_data_2.clone(),
+        signers[0].author(),
+        li2.clone(),
+        &signers[0],
+    );
+    assert_eq!(
+        pending_votes.insert_vote(&vote_data_2_author_0, Arc::clone(&validator_verifier)),
+        VoteReceptionResult::EquivocateVote
+    );
+    // A different author voting for a different result in the same round but without a round
+    // signature: VoteAdded
+    let vote_data_2_author_1 = VoteMsg::new(
+        vote_data_2.clone(),
+        signers[1].author(),
+        li2.clone(),
+        &signers[1],
+    );
+    assert_eq!(
+        pending_votes.insert_vote(&vote_data_2_author_1, Arc::clone(&validator_verifier)),
+        VoteReceptionResult::VoteAdded(1)
+    );
+    // Two votes for the ledger info form a QC
+    let vote_data_2_author_2 = VoteMsg::new(
+        vote_data_2.clone(),
+        signers[2].author(),
+        li2.clone(),
+        &signers[2],
+    );
+    match pending_votes.insert_vote(&vote_data_2_author_2, Arc::clone(&validator_verifier)) {
+        VoteReceptionResult::NewQuorumCertificate(_) => (),
+        _ => {
+            panic!("No QC formed.");
+        }
+    };
+}
```

Comment:
It might be nice to check that the struct is consistent as expected and following the expected cleanup behavior, perhaps a test consistency method.